From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Sat, 21 Sep 2024 02:50:38 +0300
Subject: [PATCH] Less limited recipe matcher

Adds the options to allow using damaged/enchanted/renamed items in the recipe matcher
used by e.g. the recipe book autofill or the crafter block.

diff --git a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
index c53d495beeda9c057efc5af1b06613be1b5fa9f4..665879a04af47617fe2007348dae94ee27f22fa3 100644
--- a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
+++ b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
@@ -150,4 +150,15 @@ public final class KiterinoConfig {
     }
     // Kiterino end - Allow server-side translatables on items
 
+    // Kiterino start - Less limited recipe matcher
+    public static boolean allowDamagedItemsInRecipeBook;
+    public static boolean allowEnchantedItemsInRecipeBook;
+    public static boolean allowRenamedItemsInRecipeBook;
+    private static void lessLimitedCraftingBook() {
+        allowDamagedItemsInRecipeBook = getBoolean(config, "recipe-matcher.allow-damaged-items", false, "Allow using damaged items when searching for items");
+        allowEnchantedItemsInRecipeBook = getBoolean(config, "recipe-matcher.allow-enchanted-items", false, "Allow using items with enchantments when searching for items");
+        allowRenamedItemsInRecipeBook = getBoolean(config, "recipe-matcher.allow-renamed-items", false, "Allow using items with custom names (i.e. renamed in anvil) when searching for items");
+    }
+    // Kiterino end - Less limited recipe matcher
+
 }
diff --git a/src/main/java/net/minecraft/world/entity/player/Inventory.java b/src/main/java/net/minecraft/world/entity/player/Inventory.java
index e7d925586f1197f50c499dc58b55cb8973d0fed1..c4eeded90f6786be9585e343669d5908e75d7fba 100644
--- a/src/main/java/net/minecraft/world/entity/player/Inventory.java
+++ b/src/main/java/net/minecraft/world/entity/player/Inventory.java
@@ -209,14 +209,14 @@ public class Inventory implements Container, Nameable {
 
             // Kiterino start - Add predicate support for recipe book
             if (validator != null) {
-                if (!itemstack1.isEmpty() && validator.test(itemstack1.asBukkitMirror()) && !itemstack1.isDamaged() && !itemstack1.isEnchanted() && !itemstack1.has(DataComponents.CUSTOM_NAME)) {
+                if (!itemstack1.isEmpty() && validator.test(itemstack1.asBukkitMirror()) && (me.sosedik.kiterino.KiterinoConfig.allowDamagedItemsInRecipeBook || !itemstack1.isDamaged()) && (me.sosedik.kiterino.KiterinoConfig.allowEnchantedItemsInRecipeBook || !itemstack1.isEnchanted()) && (me.sosedik.kiterino.KiterinoConfig.allowRenamedItemsInRecipeBook || !itemstack1.has(DataComponents.CUSTOM_NAME))) { // Kiterino - Less limited recipe matcher
                     return i;
                 }
                 continue;
             }
             // Kiterino end - Add predicate support for recipe book
 
-            if (!itemstack1.isEmpty() && ItemStack.isSameItemSameComponents(stack, itemstack1) && !itemstack1.isDamaged() && !itemstack1.isEnchanted() && !itemstack1.has(DataComponents.CUSTOM_NAME)) {
+            if (!itemstack1.isEmpty() && ItemStack.isSameItemSameComponents(stack, itemstack1) && (me.sosedik.kiterino.KiterinoConfig.allowDamagedItemsInRecipeBook || !itemstack1.isDamaged()) && (me.sosedik.kiterino.KiterinoConfig.allowEnchantedItemsInRecipeBook || !itemstack1.isEnchanted()) && (me.sosedik.kiterino.KiterinoConfig.allowRenamedItemsInRecipeBook || !itemstack1.has(DataComponents.CUSTOM_NAME))) { // Kiterino - Less limited recipe matcher
                 return i;
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/player/StackedContents.java b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
index fcabf630ce1e4949d00f485a5bff66dd1e54a277..323fa76c0cd53aae07e7a2b27609c0f8cd37abe2 100644
--- a/src/main/java/net/minecraft/world/entity/player/StackedContents.java
+++ b/src/main/java/net/minecraft/world/entity/player/StackedContents.java
@@ -26,7 +26,7 @@ public class StackedContents {
 
     public void accountSimpleStack(ItemStack stack) {
         if (this.extrasMap != null && !stack.getComponentsPatch().isEmpty() && this.extrasMap.accountStack(stack, Math.min(64, stack.getCount()))) return; // Paper - Improve exact choice recipe ingredients; max of 64 due to accountStack method below
-        if (!stack.isDamaged() && !stack.isEnchanted() && !stack.has(DataComponents.CUSTOM_NAME)) {
+        if ((me.sosedik.kiterino.KiterinoConfig.allowDamagedItemsInRecipeBook || !stack.isDamaged()) && (me.sosedik.kiterino.KiterinoConfig.allowEnchantedItemsInRecipeBook || !stack.isEnchanted()) && (me.sosedik.kiterino.KiterinoConfig.allowRenamedItemsInRecipeBook || !stack.has(DataComponents.CUSTOM_NAME))) { // Kiterino - Less limited recipe matcher
             this.accountStack(stack);
         }
     }
